src/v/kafka/server/datalake_throttle_manager.cc

```diff
@@ -0,0 +1,232 @@
+      [](datalake_throttle_manager& instance) {
+          return instance._shard_status_provider();
+      },
+      backlog_status{},
+      [](backlog_status acc, backlog_status shard_status) {
+          return acc + shard_status;
+      });
+    if (_translation_status.has_no_issues()) {
+        _last_no_issues_timestamp = clock_type::now();
+    }
+    /**
+     * Collect shard local maps and while iterating over the shards update the
+     * total backlog
+     */
+    auto shard_local_maps = co_await ssx::parallel_transform(
+      boost::irange(ss::smp::count), [this](auto shard_id) {
+          return container().invoke_on(
+            shard_id,
+            [status = _translation_status](datalake_throttle_manager& other) {
+                other._translation_status = status;
+                return std::exchange(other._shard_local_producers, {});
+            });
+      });
+
+    for (auto& shard_map : shard_local_maps) {
+        merge_producer_maps(_global_producers, shard_map);
+    }
+
+    std::erase_if(_global_producers, [this](const auto& p) {
+        return (clock_type::now() - p.second.last_datalake_produce)
+               > _producer_gc_threshold();
+    });
+}
+
+ss::future<std::chrono::milliseconds>
+datalake_throttle_manager::maybe_throttle_producer(
+  std::optional<std::string_view> client_id) {
+    // fast path, backlog is below the threshold
+
+    if (_translation_status.has_no_issues()) [[likely]] {
+        co_return no_throttling;
+    }
+
+    auto throttle_ms = co_await container().invoke_on(
+      0, [&client_id](datalake_throttle_manager& shard_0_manager) {
+          auto throttle_ms = shard_0_manager.get_producer_throttle(client_id);
+          if (throttle_ms > 0ms) {
+              vlog(
+                client_quota_log.debug,
+                "Throttling producer {} for {}ms, current "
+                "status: {}",
+                client_id,
+                throttle_ms,
+                shard_0_manager._translation_status);
+          }
+          return throttle_ms;
+      });
+    // record throttle for exposing a metric
+    _total_throttle += throttle_ms.count();
+    co_return throttle_ms;
+}
+
+std::chrono::milliseconds datalake_throttle_manager::get_producer_throttle(
+  const std::optional<std::string_view>& client_id) {
+    vassert(
+      ss::this_shard_id() == 0, "Throttle can only be calculate on shard 0");
+    if (_translation_status.has_no_issues()) [[likely]] {
+        return no_throttling;
+    }
+    auto effective_client_id = get_effective_client_id(client_id);
+
+    auto it = _global_producers.find(effective_client_id);
+    // do not throttle non datalake producers
+    if (it == _global_producers.end()) {
+        return no_throttling;
+    }
+
+    return calculate_throttle();
+}
+
+std::chrono::milliseconds
+datalake_throttle_manager::calculate_throttle() const {
+    if (_translation_status.partitions_translation_blocked > 0) {
```

Comment:
nit: shard assert here as well to assist the reader in understanding which methods are called only on shard 0? Or maybe just merge with the previous method